In [1]:
pip install pandas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 625.2 kB/s eta 0:00:001m750.8 kB/s eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 35.7 MB/s eta 0:00:00m eta 0:00:010:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.0/18.0 MB 26.8 MB/s eta 0:00:00m eta 0:00:010:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.4/345.4 kB 10.5 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [1]:
!conda install scikit-learn


Channels:
 - defaults
Platform: linux-64
Solving environment: done

## Package Plan ##

  environment location: /opt/conda

  added / updated specs:
    - scikit-learn


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    blas-1.0                   |              mkl           6 KB
    intel-openmp-2023.1.0      |   hdb19cb5_46306        17.2 MB
    joblib-1.4.0               |  py312h06a4308_0         521 KB
    libgfortran-ng-11.2.0      |       h00389a5_1          20 KB
    libgfortran5-11.2.0        |       h1234567_1         2.0 MB
    mkl-2023.1.0               |   h213fc3f_46344       171.5 MB
    mkl-service-2.4.0          |  py312h5eee18b_1          66 KB
    mkl_fft-1.3.8              |  py312h5eee18b_0         204 KB
    mkl_random-1.2.4           |  py312hdb19cb5_0         284 KB
    numpy-1.26.4               |  py312hc5e2394_0          11 KB
    numpy-base-1.26.4          |  py

In [1]:
import sklearn

In [2]:
from sklearn.feature_extraction import DictVectorizer

In [3]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

In [4]:
import pandas as pd

In [5]:
pd.__version__

'2.2.2'

In [4]:
!pip install pyarrow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.8/40.8 MB 25.7 MB/s eta 0:00:00m eta 0:00:010:01:01


In [5]:
pip install fastparquet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 5.9 MB/s eta 0:00:00m eta 0:00:010:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 21.5 MB/s eta 0:00:00m eta 0:00:010:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 316.1/316.1 kB 10.0 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


In [43]:
df = pd.read_parquet('https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2023-01.parquet')

In [6]:
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2024-01-01 00:57:55   2024-01-01 01:17:43              1.0   
1         1  2024-01-01 00:03:00   2024-01-01 00:09:36              1.0   
2         1  2024-01-01 00:17:06   2024-01-01 00:35:01              1.0   
3         1  2024-01-01 00:36:38   2024-01-01 00:44:56              1.0   
4         1  2024-01-01 00:46:51   2024-01-01 00:52:57              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           1.72         1.0                  N           186            79   
1           1.80         1.0                  N           140           236   
2           4.70         1.0                  N           236            79   
3           1.40         1.0                  N            79           211   
4           0.80         1.0                  N           211           148   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2         17.7    1.0      0.5        0.00           0.0   
1             1         10.0    3.5      0.5        3.75           0.0   
2             1         23.3    3.5      0.5        3.00           0.0   
3             1         10.0    3.5      0.5        2.00           0.0   
4             1          7.9    3.5      0.5        3.20           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  
0                    1.0         22.70                   2.5          0.0  
1                    1.0         18.75                   2.5          0.0  
2                    1.0         31.30                   2.5          0.0  
3                    1.0         17.00                   2.5          0.0  
4                    1.0         16.10                   2.5          0.0

In [45]:
len(df.columns)

19

### Q2. Computing duration
Now let's compute the duration variable. It should contain the duration of a ride in minutes.

What's the standard deviation of the trips duration in January?

32nswer
 -> Answer
42.59

52.59
62.59

In [46]:
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   
3           1.90         1.0                  N           138             7   
4           1.43         1.0                  N           107            79   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3   1.00      0.5        0.00           0.0   
1             1          7.9   1.00      0.5        4.00           0.0   
2             1         14.9   1.00      0.5       15.00           0.0   
3             1         12.1   7.25      0.5        0.00           0.0   
4             1         11.4   1.00      0.5        3.28           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    1.0         14.30                   2.5         0.00  
1                    1.0         16.90                   2.5         0.00  
2                    1.0         34.90                   2.5         0.00  
3                    1.0         20.85                   0.0         1.25  
4                    1.0         19.68                   2.5         0.00

In [11]:
# Copy the data
df_copy = df_jan_2023.copy()

In [12]:
df_copy.shape

(3066766, 19)

In [13]:
# Datetime columns are in DateTime format
df_copy['tpep_pickup_datetime'] = pd.to_datetime(df_copy['tpep_pickup_datetime'])
df_copy['tpep_dropoff_datetime'] = pd.to_datetime(df_copy['tpep_dropoff_datetime'])


In [14]:
# Calculate the duration in minutes
df_copy['duration_in_minutes'] = (df_copy['tpep_dropoff_datetime'] - df_copy['tpep_pickup_datetime']).dt.total_seconds() / 60


In [15]:
# Filter the data for trips in January
df_january = df_copy[df_copy['tpep_pickup_datetime'].dt.month == 1]

In [16]:
df_january

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1               2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2               2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3               1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4               2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   
...           ...                  ...                   ...              ...   
3066761         2  2023-01-31 23:58:34   2023-02-01 00:12:33              NaN   
3066762         2  2023-01-31 23:31:09   2023-01-31 23:50:36              NaN   
3066763         2  2023-01-31 23:01:05   2023-01-31 23:25:36              NaN   
3066764         2  2023-01-31 23:40:00   2023-01-31 23:53:00              NaN   
3066765         2  2023-01-31 23:07:32   2023-01-31 23:21:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.97         1.0                  N           161   
1                 1.10         1.0                  N            43   
2                 2.51         1.0                  N            48   
3                 1.90         1.0                  N           138   
4                 1.43         1.0                  N           107   
...                ...         ...                ...           ...   
3066761           3.05         NaN               None           107   
3066762           5.80         NaN               None           112   
3066763           4.67         NaN               None           114   
3066764           3.15         NaN               None           230   
3066765           2.85         NaN               None           262   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 141             2         9.30   1.00      0.5        0.00   
1                 237             1         7.90   1.00      0.5        4.00   
2                 238             1        14.90   1.00      0.5       15.00   
3                   7             1        12.10   7.25      0.5        0.00   
4                  79             1        11.40   1.00      0.5        3.28   
...               ...           ...          ...    ...      ...         ...   
3066761            48             0        15.80   0.00      0.5        3.96   
3066762            75             0        22.43   0.00      0.5        2.64   
3066763           239             0        17.61   0.00      0.5        5.32   
3066764            79             0        18.15   0.00      0.5        4.43   
3066765           143             0        15.97   0.00      0.5        2.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0         14.30   
1                 0.0                    1.0         16.90   
2                 0.0                    1.0         34.90   
3                 0.0                    1.0         20.85   
4                 0.0                    1.0         19.68   
...               ...                    ...           ...   
3066761           0.0                    1.0         23.76   
3066762           0.0                    1.0         29.07   
3066763           0.0                    1.0         26.93   
3066764           0.0                    1.0         26.58   
3066765           0.0                    1.0         21.97   

         congestion_surcharge  airport_fee  duration_in_minutes  
0                         2.5         0.00             8.433333  
1                         2.5         0.00             6.316667  
2                         2.5         0.00            12.750000  
3                         0.0         1.25             9.616667  
4                         2.5         0.00            10.833333  
...                       ...          ...                  ...  
3066761        

In [17]:
# Compute the standard deviation of the trip durations
std_duration = df_january['duration_in_minutes'].std()

In [18]:
print(f"The standard deviation of the trip durations in January is: {std_duration:.2f} minutes")

The standard deviation of the trip durations in January is: 42.59 minutes


#### ---------------------------------------------Q3---------------------------------------------

### Q3. Dropping outliers
Next, we need to check the distribution of the duration variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

What fraction of the records left after you dropped the outliers?


90
%
9
2%
 -> Ans95%
98% There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

What fraction of the records left after you dropped the outliers?


90
%
9
2%
 - Ans95%
98%

In [19]:
# Debugging: Check for missing values in duration_in_minutes
print("Missing values in duration_in_minutes before filtering:", df_copy['duration_in_minutes'].isna().sum())

Missing values in duration_in_minutes before filtering: 0


In [20]:
# Step 4: Examine the data before filtering
print("Data before filtering:")
print(df.head())  # Show the first few rows of the dataframe
print(f"Total number of records before filtering: {len(df_copy)}")
print("Summary statistics of duration_in_minutes:")
print(df_copy['duration_in_minutes'].describe())

Data before filtering:
   VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2024-01-01 00:57:55   2024-01-01 01:17:43              1.0   
1         1  2024-01-01 00:03:00   2024-01-01 00:09:36              1.0   
2         1  2024-01-01 00:17:06   2024-01-01 00:35:01              1.0   
3         1  2024-01-01 00:36:38   2024-01-01 00:44:56              1.0   
4         1  2024-01-01 00:46:51   2024-01-01 00:52:57              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           1.72         1.0                  N           186            79   
1           1.80         1.0                  N           140           236   
2           4.70         1.0                  N           236            79   
3           1.40         1.0                  N            79           211   
4           0.80         1.0                  N           211           148   

   payment_type  fare_amount  extra  mta_tax  tip_a

In [21]:
# Debugging: Check the number of rows before filtering
print("Number of rows after filtering:", df_copy['duration_in_minutes'].shape[0])

Number of rows after filtering: 3066766


In [53]:
# duration was between 1 and 60 minutes.
filtered_data = df_copy[(df_copy['duration_in_minutes'] >= 1) & (df_copy['duration_in_minutes'] <= 60)]


In [23]:
# Debugging: Check the number of rows after filtering
print("Number of rows after filtering:", filtered_data.shape[0])

Number of rows after filtering: 3009173


In [25]:
total_records_before = len(df_copy)

In [26]:
total_records_before

3066766

In [23]:
total_records_after = len(filtered_data)

In [24]:
total_records_after

3009173

In [25]:
# Compute the fraction of records left
fraction_left = total_records_after / total_records_before
fraction_left *100


98.1220282212598

In [26]:
print(f"Fraction of records left after dropping outliers: {fraction_left:.4f}")

Fraction of records left after dropping outliers: 0.9812


#### -------------------------------------------------Q4-------------------------------------------------------

Q4. One-hot encoding
Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

Turn tdataframeame into a list of dictionaries (remember to re-cast the ids to strings - otherwise it will label encode them)
Fit a dictionary vectorizer
Get a feature matrix from it
What's the dimensionality of this matrix (number of olumn
s)?




2




155

345

515 -> Ans

715

In [55]:
df = df[(df['duration_in_minutes'] >= 1) & (df['duration_in_minutes'] <= 60)].copy()

In [61]:
# Converted Into String
categorical = ['PULocationID', 'DOLocationID']


In [62]:
# Convert DataFrame into the list
df[categorical] = df[categorical].astype(str)


In [63]:
train_dicts = df[categorical].to_dict(orient='records')

In [64]:
len(train_dict)

3066766

In [65]:
# List of dictionaries and transform them into a feature matrix. sparse=False
dv = DictVectorizer()
X_train = dv.fit_transform(train_dict)

In [66]:
print(f'Feature matrix size: {X_train.shape}')


Feature matrix size: (3066766, 518)


#### -------------------------------------------Q4-------------------------------

### Q5. Training a model

In [59]:
target = 'duration_in_minutes'
y_train = df[target].values

In [67]:
len(y_train)

3009173

In [60]:
lr = LinearRegression()
lr.fit(X_train, y_train)



ValueError: Found input variables with inconsistent numbers of samples: [3066766, 3009173]